Dry up cost determination

vicsn opened the pull request:
## Motivation

Closes https://github.com/provableHQ/snarkVM/issues/2892
It still ignores the storage cost, we have a separate issue for that: https://github.com/provableHQ/snarkVM/issues/2891

## Test Plan

See snarkOS PR

## Related PRs

https://github.com/ProvableHQ/snarkOS/pull/3833

vicsn committed:
Introduce consensus_config_value_by_version

vicsn committed:
DRY up compute cost determination

- `{deploy, execute}_compute_cost_in_microcredits` is now re-used across snarkVM and snarkOS
- I did not want to complicate matters with yet another cost function, so I kept CostDetails a tuple.
- The `transaction_spend_limit` checks in `initialize_and_check` are moved elsewhere. To help reviewers analyse the forking risk, consider two core pieces of transaction verification logic:
  1. Verifying transmissions for incoming batch proposals
  2. Verifying linearized transmissions
Before this PR, `initialize_and_check`  was called indirectly in both paths. To ensure we keep the same guarantees, I moved the `transaction_spend_limit` checks into `{deployment, execution}_cost`, which is also called in both paths.

raychu86 commented on synthesizer/process/src/cost.rs:
```diff
@@ -118,28 +188,46 @@ pub fn deployment_cost_v1<N: Network>(
+        // Get the finalize cost.
+        let finalize_cost = cost_in_microcredits_v2(&stack, function.name())?;
+        // Check that the finalize cost does not exceed the maximum.
+        ensure!(
```
We may want to check that no program on-chain has a function that violates this new condition

raychu86 replied:
Actually, may not be relevant since you are using the first/largest transaction spend limit

vicsn replied:
Indeed, previously this logic lived in Stack:initialize_and_check, ensuring all functions were either below the old/new limit using the old/new cost function

raychu86 commented:
Seems logically fine, some breaking CI though

vicsn committed:
Check transaction_spend_limit in verify_deployment

vicsn committed:
Remove duplicate and expensive to refactor test: deployment_cost requires a deployment to verify costs, and generating many deployments in a loop would be prohibitively expensive

vicsn commented:
> Seems logically fine, some breaking CI though

Thx for the heads up, should be fixed with the latest two commits.

vicsn commented:
@raychu86 I'll only merge this once the snarkOS counterpart is also approved.

vicsn commented on synthesizer/process/src/cost.rs:
```diff
@@ -75,30 +129,46 @@ pub fn deployment_cost_v2<N: Network>(
+
     // Compute the constructor cost in microcredits.
-    let constructor_cost = constructor_cost_in_microcredits_v2(&Stack::new(process, deployment.program())?)?;
+    let constructor_cost = constructor_cost_in_microcredits_v2(&stack)?;
```
Should we check that this is below the `TRANSACTION_SPEND_LIMIT`? In either cost function?

vicsn replied:
No, the proposed design makes sense. We check the compute cost in a later stage.

vicsn committed:
Remove leftover TODO

vicsn committed:
Correct tx cost logic checks.

Before this PR, deployments' constructor costs were checked to be below
the transaction_spend_limit. This logic is only encoded in
deploy_compute_cost_in_microcredits from V10 onwards, so let's only
check from V10 onwards.

The same applies to executions and their finalize_cost.

vicsn committed:
Merge remote-tracking branch 'origin/staging' into dry_up_cost_determination
